### ISSUE TITLE
bug: unhandled overflow in backoff

### ISSUE THREAD

**Issue by vpavlin** [waku-org/nwaku#1665]
### Problem
A node randomly crashes with unhandled exception - OverflowDefect

### Impact
Not sure, potentially disruptive for production/reliable runtime, but does not seem critical

### To reproduce

1. Clone https://github.com/alrevuelta/nwaku-compose
2. Update the image version to v0.16.0
4. Run `docker compose up`
5. Observe the following error after some time

### Expected behavior
The node does not crash with unhandled exception

### Screenshots/logs
```
INF 2023-04-08 11:50:42.807+00:00 Peer reachability status                   topics="waku node" tid=1 file=waku_node.nim:257 networkReachability=Reachable confidence=1.0
INF 2023-04-08 11:50:44.196+00:00 Finished dialing multiple peers            topics="waku node peer_manager" tid=1 file=peer_manager.nim:382 successfulConns=1 attempted=4
INF 2023-04-08 11:50:44.862+00:00 Discovered peers                           topics="waku node" tid=1 file=waku_node.nim:1016 discovered=36 new=1
INF 2023-04-08 11:50:47.139+00:00 Total connections initiated                topics="waku node metrics" tid=1 file=waku_metrics.nim:67 count=6.0
INF 2023-04-08 11:50:47.139+00:00 Total messages                             topics="waku node metrics" tid=1 file=waku_metrics.nim:68 count=29737.0
INF 2023-04-08 11:50:47.139+00:00 Total store peers                          topics="waku node metrics" tid=1 file=waku_metrics.nim:69 count=0.0
INF 2023-04-08 11:50:47.139+00:00 Total peer exchange peers                  topics="waku node metrics" tid=1 file=waku_metrics.nim:70 count=0.0
INF 2023-04-08 11:50:47.139+00:00 Total lightpush peers                      topics="waku node metrics" tid=1 file=waku_metrics.nim:71 count=0.0
INF 2023-04-08 11:50:47.139+00:00 Total filter peers                         topics="waku node metrics" tid=1 file=waku_metrics.nim:72 count=0.0
INF 2023-04-08 11:50:47.139+00:00 Total active filter subscriptions          topics="waku node metrics" tid=1 file=waku_metrics.nim:73 count=46.0
INF 2023-04-08 11:50:47.139+00:00 Total errors                               topics="waku node metrics" tid=1 file=waku_metrics.nim:74 count=0.0
INF 2023-04-08 11:51:01.868+00:00 Discovered peers                           topics="waku node" tid=1 file=waku_node.nim:1016 discovered=29 new=1
/app/apps/wakunode2/wakunode2.nim(775) wakunode2
/app/vendor/nim-chronos/chronos/asyncloop.nim(1254) runForever
/app/vendor/nim-chronos/chronos/asyncloop.nim(269) poll
/app/vendor/nim-chronos/chronos/asyncfutures2.nim(369) futureContinue
/app/waku/v2/node/peer_manager/peer_manager.nim(491) relayConnectivityLoop
/app/waku/v2/node/peer_manager/peer_manager.nim(392) connectToRelayPeers
/app/vendor/nim-chronos/chronos/asyncfutures2.nim(369) futureContinue
/app/waku/v2/node/peer_manager/peer_manager.nim(406) connectToRelayPeers
/app/waku/v2/node/peer_manager/waku_peer_store.nim(62) canBeConnected
/app/vendor/nim-chronos/chronos/timer.nim(298) seconds
/app/vendor/nimbus-build-system/vendor/Nim/lib/system/integerops.nim(15) raiseOverflow
/app/vendor/nimbus-build-system/vendor/Nim/lib/system/fatal.nim(54) sysFatal
[[reraised from:
/app/apps/wakunode2/wakunode2.nim(775) wakunode2
/app/vendor/nim-chronos/chronos/asyncloop.nim(1254) runForever
/app/vendor/nim-chronos/chronos/asyncloop.nim(269) poll
/app/vendor/nim-chronos/chronos/asyncfutures2.nim(369) futureContinue
/app/waku/v2/node/peer_manager/peer_manager.nim(491) relayConnectivityLoop
/app/waku/v2/node/peer_manager/peer_manager.nim(392) connectToRelayPeers
/app/vendor/nim-chronos/chronos/asyncfutures2.nim(391) futureContinue
]]
[[reraised from:
/app/apps/wakunode2/wakunode2.nim(775) wakunode2
/app/vendor/nim-chronos/chronos/asyncloop.nim(1254) runForever
/app/vendor/nim-chronos/chronos/asyncloop.nim(269) poll
/app/vendor/nim-chronos/chronos/asyncfutures2.nim(391) futureContinue
]]
Error: unhandled exception: over- or underflow [OverflowDefect]
```

### nwaku version/commit hash
v0.16.0

**Comment by rymnc** [comment]
I think this is the offending line - 
https://github.com/waku-org/nwaku/blob/2229fa34ae008ac34cd1eefaec2747295eaa470c/waku/v2/node/peer_manager/waku_peer_store.nim#L62

There are a few other places where an underflow *may* happen, but doubt it since they're all ints
- https://github.com/waku-org/nwaku/blob/2229fa34ae008ac34cd1eefaec2747295eaa470c/waku/v2/node/peer_manager/peer_manager.nim#L317
- https://github.com/waku-org/nwaku/blob/2229fa34ae008ac34cd1eefaec2747295eaa470c/waku/v2/node/peer_manager/peer_manager.nim#L401
- https://github.com/waku-org/nwaku/blob/2229fa34ae008ac34cd1eefaec2747295eaa470c/waku/v2/node/peer_manager/peer_manager.nim#L409
- https://github.com/waku-org/nwaku/blob/2229fa34ae008ac34cd1eefaec2747295eaa470c/waku/v2/node/peer_manager/peer_manager.nim#L426
- https://github.com/waku-org/nwaku/blob/2229fa34ae008ac34cd1eefaec2747295eaa470c/waku/v2/node/peer_manager/peer_manager.nim#L433
- https://github.com/waku-org/nwaku/blob/2229fa34ae008ac34cd1eefaec2747295eaa470c/waku/v2/node/peer_manager/peer_manager.nim#L442

**Comment by rymnc** [comment]
Try this -

```nim
import chronos, std/math

let initialBackoffInSec = 120
let backoffFactor = 4
let failedAttempts = 15

discard chronos.seconds(initialBackoffInSec*(backoffFactor^(failedAttempts-1)))
``` 

errors out with a sysFatal, works OK with failedAttempts = 14

having 15 failed attempts is rare, since after the 14 attempt, this is the delay before the next attempt - `13315w1d18h8m`

Should probably update `canBeConnected`

**Comment by vpavlin** [comment]
IIRC for example in Kubernetes there is a backoff AND a limit for retries - should we also introduce a limit?

Even with failedAttempts limited to 10 it would be a year - I am not sure if it even makes sense to retry after a year

**Comment by alrevuelta** [comment]
Thanks both for narrowing down the issue. Indeed, I would say its because of a very high `failedAttempts`, that causes the `backoff` to overflow.

@vpavlin There is indeed a check here https://github.com/waku-org/nwaku/blob/master/waku/v2/node/peer_manager/peer_manager.nim#L474 where peers that errored more than `maxFailedAttempts` (configured to `MaxFailedAttempts = 5` are removed.

However, I can think of an edge case where if the peer store is never over its capacity, `prunePeerStore` will never prune anything, leaving peers with a potential very high `failedAttempts` causing this issue.

So I would say the easiest solution would be to add this check `if pm.peerStore[NumberFailedConnBook][peerId] >= pm.maxFailedAttempts` in the `canBeConnected` so that if more than x attempts, it directly returns false.

**Comment by vpavlin** [comment]
Ha! Yes, I came to the same conclusion after posting the comment and looking more into the code and this is indeed the PR I am working on:)

**Comment by alrevuelta** [comment]
While investigating this issue found this.

I didn't understand why we were having this overflow, since reaching the attempt "14" (the one causing the overflow) required 13 failed dials, and this requires months (as backoff) with the current parameters.

So having a look into the logs i found this:

<img width="1712" alt="image" src="https://user-images.githubusercontent.com/8811422/232031659-396e69bc-db49-4a4e-ae39-ffdc65c0b5b2.png">

The backoff logic works as expected, but we are calling `connectRelay` directly (that skips the backoff) (like in `connectToNodes`), which shortcuts the backoff. Perhaps in the peer exchange loop.

In other words, the backoff is only enforced in `connectToRelayPeers`. If `connectToNodes` is used directly, the backoff is skipped (but the failledAttempts is still updated). So this is why the failed dials increased that rapidly! leading to the overflow.

Will work on a fix for this on Monday, but #1670 should indeed fix the above-mentioned issue. Right now `canBeConnected` checks for the amount of failedAttempts, so we are now protected from the overflow. But the rootcause of the issue as the explained one. Otherwise we would have never reached this overflow.
